File: runners/csharp/Kikker.Billing.IntegrationTests/Helpers/MySqlComposeClient.cs
using System.Globalization;

namespace Kikker.Billing.IntegrationTests.Helpers;

public sealed class MySqlComposeClient
{
    private readonly BillingTestOptions options;
    private readonly ComposeCommandRunner compose;

    public MySqlComposeClient(BillingTestOptions options, ComposeCommandRunner compose)
    {
        this.options = options;
        this.compose = compose;
    }

    public async Task<int> QueryIntAsync(string database, string sql)
    {
        var raw = await QueryScalarAsync(database, sql);
        return int.Parse(raw, CultureInfo.InvariantCulture);
    }

    public async Task ExecuteAsync(string database, string sql)
    {
        await compose.RunAsync(
            "exec",
            "-T",
            "mysql",
            "mysql",
            "-uroot",
            $"-p{options.MySqlRootPassword}",
            $"--database={database}",
            "-e",
            sql);
    }

    public async Task<string[]> QueryTabSeparatedRowAsync(string database, string sql)
    {
        var output = await QueryScalarAsync(database, sql);
        var row = output
            .Split('\n', StringSplitOptions.RemoveEmptyEntries)
            .LastOrDefault()?
            .Split('\t');

        if (row is null || row.Length == 0)
        {
            throw new InvalidOperationException("Expected one tab-separated row from MySQL query, but query returned no data.");
        }

        return row;
    }

    public async Task<string> QueryScalarAsync(string database, string sql)
    {
        var output = await compose.RunAsync(
            "exec",
            "-T",
            "mysql",
            "mysql",
            "-uroot",
            $"-p{options.MySqlRootPassword}",
            $"--database={database}",
            "-Nse",
            sql);

        return output.Trim();
    }
}

File: runners/java/src/test/java/com/kikker/billing/integrationtests/helpers/MySqlComposeClient.java
package com.kikker.billing.integrationtests.helpers;

import com.kikker.billing.integrationtests.BillingTestOptions;

import java.io.IOException;
import java.util.Arrays;

public final class MySqlComposeClient {
    private final BillingTestOptions options;
    private final ComposeCommandRunner compose;

    public MySqlComposeClient(BillingTestOptions options, ComposeCommandRunner compose) {
        this.options = options;
        this.compose = compose;
    }

    public int queryInt(String database, String sql) throws IOException, InterruptedException {
        return Integer.parseInt(queryScalar(database, sql));
    }

    public void execute(String database, String sql) throws IOException, InterruptedException {
        compose.run(
            "exec",
            "-T",
            "mysql",
            "mysql",
            "-uroot",
            "-p" + options.mysqlRootPassword(),
            "--database=" + database,
            "-e",
            sql
        );
    }

    public String[] queryTabSeparatedRow(String database, String sql) throws IOException, InterruptedException {
        String scalarOutput = queryScalar(database, sql);
        String[] rows = scalarOutput.split("\\R");
        String row = rows.length == 0 ? "" : rows[rows.length - 1].trim();

        if (row.isEmpty()) {
            throw new IllegalStateException("Expected one tab-separated row from MySQL query, but query returned no data.");
        }

        String[] columns = row.split("\t", -1);
        if (columns.length == 0) {
            throw new IllegalStateException("Expected tab-separated columns from MySQL query, but got: " + Arrays.toString(rows));
        }

        return columns;
    }

    public String queryScalar(String database, String sql) throws IOException, InterruptedException {
        return compose.run(
            "exec",
            "-T",
            "mysql",
            "mysql",
            "-uroot",
            "-p" + options.mysqlRootPassword(),
            "--database=" + database,
            "-Nse",
            sql
        ).trim();
    }
}
